larray/tests/test_constrained_session.py

```diff
@@ -0,0 +1,703 @@
+        # new variables
+        n0 = 'first new var'
+        n1: str
+
+    declared_variable_keys += ['n1', 'n0']
+    cs = TestInheritance(a, a01, a2=a2, e=e, f=h, g=g, h=f, n1='second new var')
+    assert list(cs.keys()) == declared_variable_keys
+    # --- overriden variables ---
+    assert cs.b.equals(b2)
+    assert cs.c == 5
+    assert cs.f.equals(h)
+    assert cs.h.equals(f)
+    # --- new variables ---
+    assert cs.n0 == 'first new var'
+    assert cs.n1 == 'second new var'
+    # --- variables declared in the base class ---
+    assert cs.b024.equals(b024)
+    assert cs.a.equals(a)
+    assert cs.a2.equals(a2)
+    assert cs.anonymous.equals(anonymous)
+    assert cs.a01.equals(a01)
+    assert cs.ano01.equals(ano01)
+    assert cs.d == d
+    assert cs.e.equals(e)
+    assert cs.g.equals(g)
+
+
+@needs_pytables
+def test_init_constrainedsession_hdf():
+    cs = TestConstrainedSession(inputpath('test_session.h5'))
+    assert set(cs.keys()) == {'b', 'b024', 'a', 'a2', 'anonymous', 'a01', 'ano01', 'c', 'd', 'e', 'g', 'f', 'h'}
+
+
+def test_getitem_cs(constrainedsession):
+    test_getitem(constrainedsession)
+
+
+def test_setitem_cs(constrainedsession):
+    cs = constrainedsession
+
+    # only change values of an array -> OK
+    cs['h'] = zeros_like(h)
+
+    # trying to add an undeclared variable -> prints a warning message
+    with pytest.warns(UserWarning) as caught_warnings:
+        cs['i'] = ndtest((3, 3))
+    assert len(caught_warnings) == 1
+    assert caught_warnings[0].message.args[0] == f"'i' is not declared in '{cs.__class__.__name__}'"
+
+    # trying to set a variable with an object of different type -> should fail
+    # a) type given explicitly
+    # -> Axis
+    expected_error_msg = "instance of Axis expected"
+    with pytest.raises(TypeError) as error:
+        cs['a'] = 0
+    assert str(error.value) == expected_error_msg
+    # -> ConstrainedArray
+    expected_error_msg = "Expected object of type 'Array' or a scalar for the variable 'h' " \
+                         "but got object of type 'ndarray'"
+    with pytest.raises(TypeError) as error:
+        cs['h'] = h.data
+    assert str(error.value) == expected_error_msg
+    # b) type deduced from the given default value
+    expected_error_msg = "instance of Axis expected"
+    with pytest.raises(TypeError) as error:
+        cs['b'] = ndtest((3, 3))
+    assert str(error.value) == expected_error_msg
+
+    # trying to set a ConstrainedArray variable using a scalar -> OK
+    cs['h'] = 5
+
+    # trying to set a ConstrainedArray variable using an array with axes in different order -> OK
+    cs['h'] = h.transpose()
+
+    # trying to set a ConstrainedArray variable using an array with wrong axes -> should fail
+    # a) missing axis
+    expected_error_msg = "Array 'h' was declared with axes {a, b} but got array with axes {a} ({b} axis is missing)"
+    with pytest.raises(ValueError) as error:
+        cs['h'] = ndtest(a3)
```

Comment:
We should either allow this (and upscale/broadcast the array) or not allow scalars to fill an array.
I think we discussed this somewhere.

Reply:
I can't retrieve where we have discussed this but now you say it, it makes sense to me to allow broadcasting (= accept missing axes) to fill an array. 
But I would not agree to allow extra axes. In that case, the array is then no longer "constrained".

Reply:
I removed the error if missing axes if commit 435159541216b1d052457540fc6ee48fda12b9bf